pkg/ls/ingestdoc.go

```diff
@@ -101,18 +102,60 @@ func (i ingestCursor) getOutput() *lpg.Node {
 	return i.output[len(i.output)-1]
 }
 
-func Ingest(builder GraphBuilder, root ParsedDocNode) (*lpg.Node, error) {
+type Ingester struct {
+	mu                    sync.RWMutex
+	Schema                *Layer
+	postIngestSchemaNodes []*lpg.Node
+}
+
+func (ing *Ingester) Ingest(builder GraphBuilder, root ParsedDocNode) (*lpg.Node, error) {
 	cursor := ingestCursor{
 		input: []ParsedDocNode{root},
 	}
 	_, n, err := ingestWithCursor(builder, cursor)
 	if err != nil {
 		return n, err
 	}
-	err = builder.PostIngest(root.GetSchemaNode(), n)
+	snodes := ing.GetPostIngestSchemaNodes(root.GetSchemaNode())
+	var nodeIDMap map[string][]*lpg.Node
+	if len(snodes) > 0 {
+		nodeIDMap = GetSchemaNodeIDMap(n)
+	}
+	for _, schemaNode := range snodes {
+		builder.PostIngestSchemaNode(root.GetSchemaNode(), schemaNode, n, nodeIDMap)
+	}
 	return n, err
 }
 
+func (ing *Ingester) GetPostIngestSchemaNodes(schemaRootNode *lpg.Node) []*lpg.Node {
+	ing.mu.RLock()
+	defer ing.mu.RUnlock()
```

Comment:
You can't defer this. The next lock will block. You have to RUnlock at every possible exit point before locking again.